Sources/AST/Environment/Environment.swift

```diff
@@ -114,53 +114,97 @@ public struct Environment {
-        identifier.name != targetVariables[targetIndex].identifier.name {
-        if targetVariables[targetIndex].assignedExpression == nil {
+
+    while sourceIndex < sourceVariables.count && sourceVariables[sourceIndex].assignedExpression == nil {
+      // Check identifiers
+      if target[targetIndex].identifier != nil {
+        if target[targetIndex].identifier!.name != sourceVariables[sourceIndex].identifier.name {
           return false
-        } else {
-          targetIndex+=1
-          continue
         }
       }
-      if targetTypes[targetIndex] == type(of: source[sourceIndex].expression,
+
+      // Check types
+      if sourceTypes[sourceIndex] != type(of: target[targetIndex].expression,
                                           enclosingType: enclosingType,
                                           scopeContext: scopeContext) {
-        sourceIndex+=1
-        targetIndex+=1
-      } else {
+        // Wrong type
         return false
       }
+
+      sourceIndex += 1
+      targetIndex += 1
     }
+
+    // Check default parameters
+    while sourceIndex < sourceVariables.count && targetIndex < target.count {
+      guard let argumentIdentifier = target[targetIndex].identifier else {
+        if sourceTypes[sourceIndex] != type(of: target[targetIndex].expression,
+                                            enclosingType: enclosingType,
+                                            scopeContext: scopeContext) {
+          return false
+        }
+
+        sourceIndex += 1
+        targetIndex += 1
+        continue
+      }
+
+      while sourceIndex < sourceVariables.count &&
+            argumentIdentifier.name != sourceVariables[sourceIndex].identifier.name {
+        sourceIndex += 1
+      }
+
+      if sourceIndex == sourceVariables.count {
+        // Identifier was not found
+        return false
+      }
+
+      if sourceTypes[sourceIndex] != type(of: target[targetIndex].expression,
+                                          enclosingType: enclosingType,
+                                          scopeContext: scopeContext) {
+        // Wrong type
+        return false
+      }
+
+      sourceIndex += 1
+      targetIndex += 1
+    }
+
+    if targetIndex < target.count {
+      // Not all arguments were matches
+      return false
+    }
+
     return true
   }
 
   /// Whether to function arguments are compatible.
   ///
   /// - Parameters:
-  ///   - source: arguments of the function that the user is trying to call.
-  ///   - target: arguments of the function available in this scope.
+  ///   - source: function information of the function that the user is trying to call.
+  ///   - target: function call available in this scope.
   /// - Returns: Boolean indicating whether function arguments are compatible.
-  func areFunctionArgumentsCompatible(source: [RawType], target: [RawType], enclosingType: RawTypeIdentifier) -> Bool {
+  func areFunctionArgumentsCompatible(source: FunctionInformation,
```

Comment:
I think this function is getting a bit bloated… Can't the check for the required + default params be extracted out into a separate function? Then call this function + the extracted function wherever you are evaluating calls with defaulted arguments . We use this function for Self in #38 but I don't see much value in unnecessarily passing the function information and scope context. I'm not even sure if it's available where we're using it.